vm/proxyapp/proxyappclient.go

```diff
@@ -0,0 +1,417 @@
+// Copyright 2022 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+// Package proxyapp package implements the experimental plugins support.
+// We promise interface part will not be stable until documented.
+package proxyapp
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/rpc"
+	"net/rpc/jsonrpc"
+	"os"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/google/syzkaller/pkg/log"
+	"github.com/google/syzkaller/pkg/report"
+	"github.com/google/syzkaller/vm/proxyapp/proxyrpc"
+	"github.com/google/syzkaller/vm/vmimpl"
+)
+
+func ctor(params *proxyAppParams, env *vmimpl.Env) (vmimpl.Pool, error) {
+	subConfig, err := parseConfig(env.Config)
+	if err != nil {
+		return nil, fmt.Errorf("config parse error: %w", err)
+	}
+
+	p := &pool{
+		env:      env,
+		close:    make(chan bool, 1),
+		onClosed: make(chan error, 1),
+	}
+
+	err = p.init(params, subConfig)
+	if err != nil {
+		return nil, fmt.Errorf("can't initialize pool: %w", err)
+	}
+
+	go func() {
+		for {
+			select {
+			case <-p.close:
+				p.mu.Lock()
+				if p.proxy != nil {
+					p.proxy.terminate()
+					<-p.proxy.onTerminated
+				}
+				p.proxy = nil
+				p.onClosed <- nil
+				p.mu.Unlock()
+				return
+			case <-p.proxy.onTerminated:
+			}
+			p.mu.Lock()
+			time.Sleep(params.InitRetryDelay)
+			p.init(params, subConfig)
+			p.mu.Unlock()
+		}
+	}()
+
+	return p, nil
+}
+
+type pool struct {
+	mu       sync.Mutex
+	env      *vmimpl.Env
+	proxy    *ProxyApp
+	count    int
+	close    chan bool
+	onClosed chan error
+}
+
+func (p *pool) init(params *proxyAppParams, cfg *Config) error {
+	var err error
+	p.proxy, err = runProxyApp(params, cfg.Command)
+	if err != nil {
+		return fmt.Errorf("failed to run ProxyApp: %w", err)
+	}
+
+	count, err := p.proxy.CreatePool(string(cfg.ProxyAppConfig), p.env.Debug)
+	if err != nil || count == 0 || (p.count != 0 && p.count != count) {
+		if err == nil {
+			err = fmt.Errorf("wrong pool size %v, prev was %v", count, p.count)
+		}
+
+		p.proxy.terminate()
+		<-p.proxy.onTerminated
+		p.proxy = nil
+
+		return fmt.Errorf("failed to construct pool: %w", err)
+	}
+
+	if p.count == 0 {
+		p.count = count
+	}
+	return nil
+}
+
+func (p *pool) Count() int {
+	return p.count
+}
+
+func (p *pool) Create(workdir string, index int) (vmimpl.Instance, error) {
+	p.mu.Lock()
+	proxy := p.proxy
+	p.mu.Unlock()
+
+	if proxy == nil {
+		return nil, fmt.Errorf("can't create instance using nil pool")
+	}
+
+	return proxy.CreateInstance(workdir, index)
+}
+
+// Close is not used now. Its support require wide code changes.
+// TODO: support the pool cleanup on syz-manager level.
+func (p *pool) Close() error {
+	close(p.close)
+	return <-p.onClosed
+}
+
+// proxyAppInterface is used for tests only. But it is here to let go:generate work.
+// nolint:deadcode, unused
```

Comment:
There are 2 useful ways to use it in the code and also it's useful for documentation. This is basically the interface users need to implement when working on proxies.

1. It can be used to verify that a type implements this interface. E.g. you can add:
```
package proxyapp

//go:generate ../../tools/mockery.sh --name subProcessCmd --exported
//go:generate ../../tools/mockery.sh --name proxyAppInterface --exported
...
var (
	_ subProcessCmd     = &mocks.SubProcessCmd{}
	_ proxyAppInterface = &mocks.ProxyAppInterface{}
)
```
This is a bit of tautology since we auto-generated implementations from the interfaces (unless mockery introduces bugs in future), but is more useful for real implementations.

2. Currently go test output is polluted with tons of these messages:
```
2022/09/24 13:07:32 rpc.Register: method "AssertCalled" has 4 input parameters; needs exactly three
2022/09/24 13:07:32 rpc.Register: method "AssertExpectations" has 2 input parameters; needs exactly three
2022/09/24 13:07:32 rpc.Register: method "AssertNotCalled" has 4 input parameters; needs exactly three
2022/09/24 13:07:32 rpc.Register: method "AssertNumberOfCalls" has 4 input parameters; needs exactly three
2022/09/24 13:07:32 rpc.Register: method "Called" has 2 input parameters; needs exactly three
2022/09/24 13:07:32 rpc.Register: method "IsMethodCallable" has 4 input parameters; needs exactly three
2022/09/24 13:07:32 rpc.Register: reply type of method "MethodCalled" is not a pointer: "[]interface {}"
2022/09/24 13:07:32 rpc.Register: reply type of method "On" is not a pointer: "[]interface {}"
2022/09/24 13:07:32 rpc.Register: method "String" has 1 input parameters; needs exactly three
2022/09/24 13:07:32 rpc.Register: method "Test" has 2 input parameters; needs exactly three
2022/09/24 13:07:32 rpc.Register: method "TestData" has 1 input parameters; needs exactly three
...
2022/09/24 13:07:32 rpc.Register: method "AssertCalled" has 4 input parameters; needs exactly three
2022/09/24 13:07:32 rpc.Register: method "AssertExpectations" has 2 input parameters; needs exactly three
2022/09/24 13:07:32 rpc.Register: method "AssertNotCalled" has 4 input parameters; needs exactly three
2022/09/24 13:07:32 rpc.Register: method "AssertNumberOfCalls" has 4 input parameters; needs exactly three
2022/09/24 13:07:32 rpc.Register: method "Called" has 2 input parameters; needs exactly three
2022/09/24 13:07:32 rpc.Register: method "IsMethodCallable" has 4 input parameters; needs exactly three
2022/09/24 13:07:32 rpc.Register: reply type of method "MethodCalled" is not a pointer: "[]interface {}"
2022/09/24 13:07:32 rpc.Register: reply type of method "On" is not a pointer: "[]interface {}"
2022/09/24 13:07:32 rpc.Register: method "String" has 1 input parameters; needs exactly three
2022/09/24 13:07:32 rpc.Register: method "Test" has 2 input parameters; needs exactly three
2022/09/24 13:07:32 rpc.Register: method "TestData" has 1 input parameters; needs exactly three
...
2022/09/24 13:07:32 rpc.Register: method "AssertCalled" has 4 input parameters; needs exactly three
2022/09/24 13:07:32 rpc.Register: method "AssertExpectations" has 2 input parameters; needs exactly three
2022/09/24 13:07:32 rpc.Register: method "AssertNotCalled" has 4 input parameters; needs exactly three
2022/09/24 13:07:32 rpc.Register: method "AssertNumberOfCalls" has 4 input parameters; needs exactly three
2022/09/24 13:07:32 rpc.Register: method "Called" has 2 input parameters; needs exactly three
2022/09/24 13:07:32 rpc.Register: method "IsMethodCallable" has 4 input parameters; needs exactly three
2022/09/24 13:07:32 rpc.Register: reply type of method "MethodCalled" is not a pointer: "[]interface {}"
2022/09/24 13:07:32 rpc.Register: reply type of method "On" is not a pointer: "[]interface {}"
2022/09/24 13:07:32 rpc.Register: method "String" has 1 input parameters; needs exactly three
2022/09/24 13:07:32 rpc.Register: method "Test" has 2 input parameters; needs exactly three
2022/09/24 13:07:32 rpc.Register: method "TestData" has 1 input parameters; needs exactly three
```

We can use this interface to prevent these messages by doing:
```
	server.RegisterName("ProxyVM", struct{ proxyAppInterface }{handler})
```

I would move this interface into `proxyrpc` package, put it at the top and document as "This is the interface you need to implement...".

Reply:
$ go test ./vm/proxyapp/
ok      github.com/google/syzkaller/vm/proxyapp 1.762s

How to reproduce these messages?

Reply:
Both applied. Thanks.